File: app/src/main/java/com/tomaszrykala/dorefindmi/things/controller/abcbuttons/AbcButtonsController.java
package com.tomaszrykala.dorefindmi.things.controller.abcbuttons;

import com.tomaszrykala.dorefindmi.model.AbcButton;
import com.tomaszrykala.dorefindmi.things.controller.BaseController;
import com.tomaszrykala.dorefindmi.things.supplier.abcbuttons.AbcButtonsSupplier;

public class AbcButtonsController implements BaseController, AbcButtonsSupplier.Listener {

    private final AbcButtonsSupplier supplier;
    private AbcButton.Listener listener;

    private boolean isEnabled;
    private AbcButton lastPressed;

    public void setListener(AbcButton.Listener listener) {
        this.listener = listener;
    }

    public void enable() {
        isEnabled = true;
    }

    public void disable() {
        isEnabled = false;
    }

    public boolean isEnabled() {
        return isEnabled;
    }

    public boolean isLastPressed(AbcButton abcButton) {
        return lastPressed == abcButton;
    }

    public void setLastPressed(AbcButton abcButton) {
        lastPressed = abcButton;
    }

    public boolean hasLastPressed() {
        return lastPressed != null;
    }

    public AbcButtonsController(AbcButtonsSupplier abcButtonsSupplier) {
        supplier = abcButtonsSupplier;
        supplier.setListener(this);
        supplier.init();
    }

    @Override public void onButtonEvent(AbcButton abcButton, boolean pressed) {
        if (pressed) {
            switch (abcButton) {
                case A:
                    listener.onAbcButton(AbcButton.A);
                    break;
                case B:
                    listener.onAbcButton(AbcButton.B);
                    break;
                case C:
                    listener.onAbcButton(AbcButton.C);
                    break;
            }
        }
    }

    @Override public void close() throws Exception {
        supplier.close();
    }

}

File: app/src/main/java/com/tomaszrykala/dorefindmi/things/controller/abcbuttons/AbcButtonsController.kt
package com.tomaszrykala.dorefindmi.things.controller.abcbuttons

import android.support.annotation.VisibleForTesting
import com.tomaszrykala.dorefindmi.domain.AbcButton
import com.tomaszrykala.dorefindmi.things.controller.BaseController
import com.tomaszrykala.dorefindmi.things.supplier.abcbuttons.AbcButtonsSupplier

class AbcButtonsController(private val supplier: AbcButtonsSupplier) : BaseController, AbcButtonsSupplier.Listener {

    private lateinit var listener: AbcButton.Listener
    private var lastPressed: AbcButton? = null

    init {
        supplier.setListener(this)
    }

    var isEnabled: Boolean = false
        private set

    fun setListener(listener: AbcButton.Listener) {
        this.listener = listener
    }

    fun enable() {
        isEnabled = true
    }

    fun disable() {
        isEnabled = false
    }

    fun setLastPressed(abcButton: AbcButton?) {
        lastPressed = abcButton
    }

    override fun onButtonEvent(abcButton: AbcButton, pressed: Boolean) {
        if (pressed) {
            listener.onAbcButton(abcButton)
        }
    }

    @Throws(Exception::class)
    override fun close() {
        supplier.close()
    }

    @VisibleForTesting
    fun isLastPressed(abcButton: AbcButton): Boolean = lastPressed === abcButton

    @VisibleForTesting
    fun hasLastPressed(): Boolean = lastPressed != null
}
